### PR TITLE
make sparsemax nan and infinity safe

### PR DESCRIPTION
**Pull request by AndreasMadsen**
logits that are -inf will be given 0 probability and logits that are
inf will result in a nan output. Likewise if all logits are -inf the
output will also be nan.

This is done by using where operators, mostly because `0 * inf = nan`
and `x/0 = sign(x) inf` following the IEEE 754 standard. However these
results are not mathematically correct in the context of the sparsemax
algorithm.

Fixes: https://github.com/tensorflow/tensorflow/issues/15564

### CODE REVIEW

File: tensorflow/contrib/sparsemax/python/ops/sparsemax_loss.py [hunk 1]
```diff
@@ -47,14 +47,34 @@ def sparsemax_loss(logits, sparsemax, labels, name=None):
     sparsemax = ops.convert_to_tensor(sparsemax, name="sparsemax")
     labels = ops.convert_to_tensor(labels, name="labels")
 
-    shifted_logits = logits - \
-        math_ops.reduce_mean(logits, axis=1)[:, array_ops.newaxis]
+    # In the paper, they call the logits z
```
Comment (rmlarsen) [inline tensorflow/contrib/sparsemax/python/ops/sparsemax_loss.py]: Missing period at the end.

File: tensorflow/contrib/sparsemax/python/ops/sparsemax.py [hunk 2]
```diff
@@ -49,7 +49,14 @@ def sparsemax(logits, name=None):
     obs = array_ops.shape(logits)[0]
     dims = array_ops.shape(logits)[1]
 
-    z = logits - math_ops.reduce_mean(logits, axis=1)[:, array_ops.newaxis]
+    # In the paper, they call the logits z
```
Comment (rmlarsen) [inline tensorflow/contrib/sparsemax/python/ops/sparsemax.py]: Missing period at the end of sentence.

**Review by rmlarsen — CHANGES_REQUESTED** [review]
A few nits. Otherwise looks good to me.

**Comment by AndreasMadsen** [comment]
@rmlarsen Thanks. Please take another look :)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by martinwicke — DISMISSED** [review]

**Comment by martinwicke** [comment]
Adding API review label to discuss moving sparsemax into core.

**Comment by AndreasMadsen** [comment]
@martinwicke I will go on vacation for a week. I will fix this when I get back.

**Comment by AndreasMadsen** [comment]
@martinwicke I think I fixed the issues. Can you run the CI again?

**Comment by rmlarsen** [comment]
@AndreasMadsen Sorry about the delay. merging now.

**Review by rmlarsen — APPROVED** [review]
